src/linear_algebra/trace.lean

```diff
@@ -115,6 +116,62 @@ end
 section
 variables (R : Type u) [field R] {M : Type v} [add_comm_group M] [module R M]
 
+/-- The trace of a linear map correspond to the contraction pairing under the isomorphism
+ `End(M) ≃ M* ⊗ M`-/
+@[simp] lemma trace_eq_contract [finite_dimensional R M] :
+  (linear_map.trace R M) ∘ₗ (dual_tensor_hom R M M) = contract_left R M :=
+begin
+  have b := fin_basis R M,
+  apply basis.ext (basis.tensor_product (b.dual_basis) b),
+  rintros ⟨i, j⟩,
+  simp only [function.comp_app, basis.tensor_product_apply, basis.coe_dual_basis, coe_comp],
+  rw [trace_eq_matrix_trace R b, to_matrix_dual_tensor_hom],
+  by_cases hij : i = j,
+  { rw [hij, matrix.std_basis_matrix.trace_same], simp },
+  rw matrix.std_basis_matrix.trace_zero j i (1:R) hij,
+  simp [finsupp.single_eq_pi_single, hij],
+end
+
+section
+
+lemma eq_comp_symm {α β γ : Type*} (e : α ≃ β) (f : β → γ) (g : α → γ) :
+  f = g ∘ e.symm ↔ f ∘ e = g := by {split; intro h; ext, simp [h], simp [←h]}
+
+lemma comp_symm_eq {α β γ : Type*} (e : α ≃ β) (f : β → γ) (g : α → γ) :
+  g ∘ e.symm = f ↔ g = f ∘ e := by {split; intro h; ext, simp [←h], simp [h]}
+
+lemma eq_symm_comp {α β γ : Type*} (e : α ≃ β) (f : γ → α) (g : γ → β) :
+  f = e.symm ∘ g ↔ e ∘ f = g := by {split; intro h; ext, simp [h], simp [←h]}
+
+lemma symm_comp_eq {α β γ : Type*} (e : α ≃ β) (f : γ → α) (g : γ → β) :
+  e.symm ∘ g = f ↔ g = e ∘ f := by {split; intro h; ext, simp [←h], simp [h]}
```

Comment:
These belong in `data/equiv/basic`, probably right below `equiv.eq_symm_apply`. You should be able to prove these with a one-line proof along the lines of `ext e.eq_symm_apply`.

Reply:
I think it would probably be worth a separate PR for these 8 lemmas, ideally with the same lemmas for `add_equiv` too